yarpl/test/Observable_test.cpp

```diff
@@ -470,3 +468,85 @@ TEST(Observable, CancelReleasesObjects) {
   auto collector = make_ref<CollectingObserver<int>>();
   observable->subscribe(collector);
 }
+
+class InfiniteAsyncTestOperator : public ObservableOperator<int, int> {
+ public:
+  InfiniteAsyncTestOperator(Reference<Observable<int>> upstream, testing::MockFunction<void()>& checkpoint)
+      : ObservableOperator<int, int>(std::move(upstream)), checkpoint_(checkpoint) {}
+
+  Reference<Subscription> subscribe(Reference<Observer<int>> observer) override {
+    auto subscription = make_ref<TestSubscription>(get_ref(this), std::move(observer), checkpoint_);
+    ObservableOperator<int, int>::upstream_->subscribe(
+        // Note: implicit cast to a reference to a observer.
+        subscription
+        );
+    return subscription;
+  }
+
+ private:
+  class TestSubscription : public ObservableOperator<int, int>::OperatorSubscription {
+    using Super = typename ObservableOperator<int, int>::OperatorSubscription;
+
+    ~TestSubscription() {
+      t_.join();
+    }
+
+   public:
+    TestSubscription(
+        Reference<Observable<int>> observable,
+        Reference<Observer<int>> observer,
+        testing::MockFunction<void()>& checkpoint)
+        : Super(std::move(observable), std::move(observer)), checkpoint_(checkpoint) {
+
+    }
+
+    void onSubscribe(yarpl::Reference<Subscription> subscription) override {
+      Super::onSubscribe(std::move(subscription));
+      t_ = std::thread([this](){
+        while (!isCancelled()) {
+          Super::observerOnNext(1);
+        }
+        checkpoint_.Call();
+      });
+    }
+    void onNext(int value) override {
+    }
+
+    std::thread t_;
+    testing::MockFunction<void()>& checkpoint_;
+  };
+
+  testing::MockFunction<void()>& checkpoint_;
+};
+
+TEST(Observable, CancelSubscriptionChain) {
+  std::atomic_int emitted{0};
+  testing::MockFunction<void()> checkpoint;
+  testing::MockFunction<void()> checkpoint2;
+  testing::MockFunction<void()> checkpoint3;
+  std::thread t;
+  auto infinite1 = Observable<int>::create([&](Reference<Observer<int>> obs) {
+    EXPECT_CALL(checkpoint, Call()).Times(1);
+    EXPECT_CALL(checkpoint2, Call()).Times(1);
+    EXPECT_CALL(checkpoint3, Call()).Times(1);
+    t = std::thread([obs, &emitted, &checkpoint](){
+      while (!obs->isUnsubscribed()) {
+        ++emitted;
+        obs->onNext(0);
+      }
+      checkpoint.Call();
+    });
+  });
+  auto infinite2 = infinite1->skip(1)->skip(1);
+  auto test1 = make_ref<InfiniteAsyncTestOperator>(infinite2, checkpoint2);
+  auto test2 = make_ref<InfiniteAsyncTestOperator>(test1->skip(1), checkpoint3);
+  auto skip = test2->skip(8);
+
+  auto subscription = skip->subscribe([](int){});
+  while(emitted < 1000);
```

Comment:
nit: better to have a timed wait, in case of failure, we can early notice instead of waiting infinitely.

Reply:
yeah, good idea, I will fix this.